src/switchroot/ostree-prepare-root.c

```diff
@@ -259,41 +308,85 @@ main(int argc, char *argv[])
       if (mount ("usr", "usr", NULL, MS_BIND, NULL) < 0)
 	{
 	  perrorv ("failed to bind mount (class:readonly) /usr");
-	  exit (EXIT_FAILURE);
+	  die ();
 	}
       if (mount ("usr", "usr", NULL, MS_BIND | MS_REMOUNT | MS_RDONLY, NULL) < 0)
 	{
 	  perrorv ("failed to bind mount (class:readonly) /usr");
-	  exit (EXIT_FAILURE);
+	  die ();
 	}
     }
 
-  if (fchdir (orig_cwd_dfd) < 0)
+  touch_run_ostree ();
+
+  if (strcmp(root_mountpoint, "/") == 0)
     {
-      perrorv ("failed to chdir to orig root");
-      exit (EXIT_FAILURE);
+      /* pivot_root rotates two mount points around.  In this instance . (the
+       * deploy location) becomes / and the existing / becomes /sysroot.  We
+       * have to use pivot_root rather than mount --move in this instance
+       * because our deploy location is mounted as a subdirectory of the real
+       * sysroot, so moving sysroot would also move the deploy location.   In
+       * reality attempting mount --move would fail with EBUSY. */
+      if (pivot_root (".", "sysroot") < 0)
+        {
+          perrorv ("failed to pivot_root to deployment");
+          die ();
+        }
     }
-  (void) close (orig_cwd_dfd);
+  else
+    {
+      /* In this instance typically we have our ready made-up up root at
+       * /sysroot/ostree/deploy/.../ (deploy_path) and the real rootfs at
+       * /sysroot (root_mountpoint).  We want to end up with our made-up root at
+       * /sysroot/ and the real rootfs under /sysroot/sysroot as systemd will be
+       * responsible for moving /sysroot to /.
+       *
+       * We need to do this in 3 moves to avoid trying to move /sysroot under
+       * itself:
+       *
+       * 1. /sysroot/ostree/deploy/... -> /sysroot.tmp
+       * 2. /sysroot -> /sysroot.tmp/sysroot
+       * 3. /sysroot.tmp -> /sysroot
+       */
+      if (mkdir ("/sysroot.tmp", 0755) < 0)
+        {
+          perrorv ("couldn't create temporary sysroot /sysroot.tmp: ");
+          die ();
+        }
 
-  touch_run_ostree ();
+      if (mount (deploy_path, "/sysroot.tmp", NULL, MS_MOVE, NULL) < 0)
+        {
+          perrorv ("failed to MS_MOVE '%s' to '/sysroot.tmp'", deploy_path);
+          die ();
+        }
 
-  /* Move physical root to $deployment/sysroot */
-  snprintf (destpath, sizeof(destpath), "%s/sysroot", newroot);
-  if (mount (root_mountpoint, destpath, NULL, MS_MOVE, NULL) < 0)
-    {
-      perrorv ("Failed to MS_MOVE %s to '%s'", root_mountpoint, destpath);
-      exit (EXIT_FAILURE);
+      if (chdir ("/sysroot.tmp") < 0)
```

Comment:
I'm not sure if this `chdir` is necessary.  I haven't tried this yet.  I'd like to remove it if it works without.